torch/distributions.py

```diff
@@ -262,3 +262,107 @@ def log_prob(self, value):
         return (self.alpha * torch.log(self.beta) +
                 (self.alpha - 1) * torch.log(value) -
                 self.beta * value - torch.lgamma(self.alpha))
+
+
+def _dirichlet_sample_nograd(alpha):
+    gammas = torch._C._standard_gamma(alpha)
+    return gammas / gammas.sum(-1, True)
+
+
+class _Dirichlet(Function):
+    @staticmethod
+    def forward(ctx, alpha):
+        x = _dirichlet_sample_nograd(alpha)
+        ctx.save_for_backward(x, alpha)
+        return x
+
+    @staticmethod
+    @once_differentiable
+    def backward(ctx, grad_output):
+        x, alpha = ctx.saved_tensors
+        total = alpha.sum(-1, True).expand_as(alpha)
+        grad = torch._C._dirichlet_grad(x, alpha, total)
+        return grad_output * grad
+
+
+class Dirichlet(Distribution):
+    r"""
+    Creates a Dirichlet distribution parameterized by concentration `alpha`.
+
+    Example::
+
+        >>> m = Dirichlet(torch.Tensor([0.5, 0.5]))
+        >>> m.sample()  # Dirichlet distributed with concentrarion alpha
+         0.1046
+         0.8954
+        [torch.FloatTensor of size 2]
+
+    Args:
+        alpha (Tensor or Variable): concentration parameter of the distribution
+    """
+    has_rsample = True
+
+    def __init__(self, alpha):
+        self.alpha = alpha
+
+    def _sample(self, alpha):
```

Comment:
Is this needed? It seems to me that `sample_n` could just call `sample` if you moved this code there

Reply:
This helper is needed until #4100 is merged. Currently `.sample_n()` would not be able to change the shape of `self.alpha` for use inside `.sample()`. Once `.sample()` can accept an `n` or `sample_shape` parameter, we can combine all these methods into a single `.sample(sample_shape)` method.